src/app/gui/queryresults/queryresultsservice.js

```diff
@@ -1256,14 +1258,31 @@ proto.getVectorLayerFeaturesFromQueryRequest = function(vectorLayer, query={}){
 };
 
 proto._addVectorLayersDataToQueryResponse = function() {
-  this.onbefore('setQueryResponse', (queryResponse, options={}) => {
-    const {query={}} = queryResponse;
-    const {add=false}= options;
-    !add && this._vectorLayers.forEach(vectorLayer => {
-      const responseObj = this.getVectorLayerFeaturesFromQueryRequest(vectorLayer, query);
-      if(!queryResponse.data) queryResponse.data = [];
-      queryResponse.data.push(responseObj);
-    })
+  this.onbefore('setQueryResponse', (queryResponse, options = {}) => {
+    const catalogService = GUI.getService('catalog');
+
+    // sanity checks
+    if (!queryResponse.data)  queryResponse.data  = [];
+    if (!queryResponse.query) queryResponse.query = {external:{add: true, selected:false}};
+    // skip when add response to current results using addLayerFeaturesToResultsAction or external false
+    if (options.add || false === queryResponse.query.external.add) {
+      return;
+    }
+
+    // add vector layers to query response
+    this._vectorLayers
+      .forEach(layer => {
+        // always visible
+        if (layer.getVisible()) {
+          if (true === queryResponse.query.external.filter.SELECTED && !catalogService.isExternalLayerSelected({ id: layer.get('id'), type: 'vector' })) {
+            return;
+          } else if (false === queryResponse.query.external.filter.SELECTED && catalogService.isExternalLayerSelected({ id: layer.get('id'), type: 'vector'})) {
+            return;
+          } else {
+            queryResponse.data.push(this.getVectorLayerFeaturesFromQueryRequest(layer, queryResponse.query));
+          }
+        }
+      });
   });
```

Comment:
here it can be a bit difficult to follow the logic of the code

there are many return conditions inside a loop whose intent is to push new elements into an array.

Try to check if it can be equivalent re-written like below:


```suggestion
    this._vectorLayers
      .forEach(layer => {
        const isExternalLayerSelected = catalogService.isExternalLayerSelected({ id: layer.get('id'), type: 'vector' });
        const isExternalLayerFilter   = queryResponse.query.external.filter.SELECTED;
        if (
          layer.getVisible() &&
          (false === isExternalLayerFilter || isExternalLayerSelected) &&
          (true === isExternalLayerFilter || !isExternalLayerSelected)
          ) {
            queryResponse.data.push(this.getVectorLayerFeaturesFromQueryRequest(layer, queryResponse.query));
        }
      });
  });
```